lang/ir/src/ir/item_impl/callable.rs

```diff
@@ -0,0 +1,557 @@
+/// - Now we take the BLAKE-2 hash of `C` which results in 32 bytes of output
+///   and take the first 4 bytes that are returned in order as the composed
+///   selector.
+///
+/// # Examples
+///
+/// ## Overriding the composed selector
+///
+/// Given
+///
+/// ```no_compile
+/// impl MyStorage {
+///     #[ink(message, selector = "0xDEADBEEF")]
+///     fn my_message(&self) {}
+/// }
+/// ```
+///
+/// ... then the selector of `my_message` is simply `0xDEADBEEF` since it overrides
+/// the composed selector.
+///
+/// ## Inherent implementation block
+///
+/// Given
+///
+/// ```no_compile
+/// impl MyStorage {
+///     #[ink(message)]
+///     fn my_message(&self) {}
+/// }
+/// ```
+///
+/// ... then the selector of `my_message` is composed such as:
+/// ```no_compile
+/// BLAKE2("my_message".to_string().as_bytes())[0..4]
+/// ```
+///
+/// ## Trait implementation block
+///
+/// Given
+///
+/// ```no_compile
+/// impl MyTrait for MyStorage {
+///     #[ink(message)]
+///     fn my_message(&self) {}
+/// }
+/// ```
+///
+/// ... then the selector of `my_message` is composed such as:
+/// ```no_compile
+/// BLAKE2("MyTrait::my_message".to_string().as_bytes())[0..4]
+/// ```
+///
+/// ## Using full path for trait
+///
+/// Given
+///
+/// ```no_compile
+/// impl ::my_full::long_path::MyTrait for MyStorage {
+///     #[ink(message)]
+///     fn my_message(&self) {}
+/// }
+/// ```
+///
+/// ... then the selector of `my_message` is composed such as:
+/// ```no_compile
+/// BLAKE2("::my_full::long_path::MyTrait::my_message".to_string().as_bytes())[0..4]
+/// ```
+///
+/// ## Using a namespace
+///
+/// Given
+///
+/// ```no_compile
+/// #[ink(namespace = "my_namespace")]
+/// impl MyTrait for MyStorage {
+///     #[ink(message)]
+///     fn my_message(&self) {}
+/// }
+/// ```
+///
+/// ... then the selector of `my_message` is composed such as:
+/// ```no_compile
+/// BLAKE2("my_namespace::MyTrait::my_message".to_string().as_bytes())[0..4]
+/// ```
+///
+/// ## Note
+///
+/// All above examples work similarly for ink! constructors interchangeably.
+///
+/// ## Usage Recommendations
+///
+/// These recommendation mainly apply to trait implementation blocks:
+///
+/// - The recommandation by the ink! team is to use the full-path approach
+/// wherever possible; OR import the trait and use only its identifier with
+/// an additional namespace if required to disambiguate selectors.
+/// - Try not to intermix the above recommendations.
+/// - Avoid directly setting the selector of an ink! message or constuctor.
+///   Only do this if nothing else helps and you need a very specific selector,
+///   e.g. in case of backwards compatibility.
+/// - Do not use the namespace unless required to disambiguate.
+pub fn compose_selector<C>(item_impl: &ir::ItemImpl, callable: &C) -> ir::Selector
+where
+    C: Callable,
+{
+    if let Some(selector) = callable.user_provided_selector() {
+        return *selector
```

Comment:
Could warn if providing both `selector` and `namespace`, that `namespace` will be ignored.

Reply:
Sadly as of today we cannot emit compile time warnings. :(
https://internals.rust-lang.org/t/pre-rfc-add-compile-warning-macro/9370